### PR TITLE
Support multiple router calls and address issue #2175

### LINKED ISSUE #2175: Only the Last Router Triggers Events, Others Are Ignored
### Description

The problem is that even if all the routers are triggered and now need to trigger the event afterward, only the last router will activate it, while the rest won't route anything, even if their conditions are true.
```
@start()
def scan_medical(self):
    print("Scanning medical data")
    result = (
        MedicalScannerCrew()
        .crew()
        .kickoff(inputs={"data": str(self.state.patient_data)})
    )
    print("Scan result: ", result)
    self.state.total_token_usage += result.token_usage.total_tokens
    print("Medical data obtained", result)
    self.state.scan_medical = result.raw
    print("Saving medical scan data")
    with open("medical_scan.md", "w") as f:
        f.write(result.raw)

@router(scan_medical)
def diagnose_conditions(self):
    print("Diagnosing medical conditions")
    result = MedicalDiagnosisCrew().crew().kickoff(inputs={"data": self.state.scan_medical})
    self.state.diagnosed_conditions = result.raw
    print("Diagnosed conditions: ", self.state.diagnosed_conditions)

@router(diagnose_conditions)
def diabetes_router(self):
    print("Diabetes Router")
    if "D" in self.state.diagnosed_conditions:
        print("Diabetes detected")
        return "diabetes"

@listen("diabetes")
def diabetes_analysis(self):
    print("Performing detailed diabetes analysis")

@router(diagnose_conditions)
def hypertension_router(self):
    print("Hypertension Router")
    if "H" in self.state.diagnosed_conditions:
        print("Hypertension detected")
        return "hypertension"

@listen("hypertension")
def hypertension_analysis(self):
    print("Performing detailed hypertension analysis")

@router(diagnose_conditions)
def anemia_router(self):
    print("Anemia Router")
    if "A" in self.state.diagnosed_conditions:
        print("Anemia detected")
        return "anemia"

@listen("anemia")
def anemia_analysis(self):
    print("Performing detailed anemia analysis")
```
 only anemia_analysis work but others no even if all the conditions true
### Steps to Reproduce

use the same code

### Expected behavior

all the routers should activate the next node if the condition is true

### Screenshots/Code snippets

in description

### Operating System

Ubuntu 20.04

### Python Version

3.11

### crewAI Version

0.100.1

### crewAI Tools Version

0.32.0

### Virtual Environment

Venv

### Evidence

see description

### Possible Solution

none

### Additional context

none

### PR DESCRIPTION
**Pull request by bhancockio**
Closes #2175

### CODE REVIEW

**Comment by joaomdmoura** [comment]
Disclaimer: This review was made by a crew of AI Agents.

# Code Review Comment for PR #2231

## Overview
This pull request introduces substantial enhancements to the flow execution logic to accommodate multiple router calls, aligning with the requirements from issue #2175. The changes span both functionality and test coverage improvements across the files `src/crewai/flow/flow.py` and `tests/flow_test.py`.

## src/crewai/flow/flow.py

### Positive Aspects
- **Multiple Router Handling**: The improved logic allows for better handling of multiple router results, making the flow more adaptable to complex routing scenarios.
- **Documentation**: The clear comments provided in the code clarify the intention behind each method, promoting better understanding.
- **Asynchronous Patterns**: Proper use of async/await patterns is evident, enhancing the performance of the application in concurrent environments.
- **Error Handling**: The error handling features are robust, which is vital for maintaining application integrity during execution.

### Suggested Improvements

1. **Code Clarity in Router Results Collection**:
   - **Current Implementation**: The collection of results is currently entangled with execution logic which might hinder readability.
   - **Improvement**: Consider refactoring the result collection into its own function for clarity. For example:
     ```python
     async def _collect_router_results(self, trigger_method: str, result: Any) -> List[str]:
         # Proposed logic encapsulated in this function
     ```

2. **Reducing Redundant Null Checks**:
   - **Current Implementation**: Repeated checks for `None` values throughout the code can lead to unnecessary complexity.
   - **Improvement**: Consolidate these checks for more straightforward logic flow:
     ```python
     async def _execute_listeners(self, trigger_method: str, result: Any) -> None:
         valid_triggers = [t for t in [trigger_method] + router_results if t is not None]
     ```

3. **Implement Type Hinting**:
   - **Current Implementation**: Lack of type hints can impede maintainability and clarity.
   - **Improvement**: Introduce type hints to all method signatures, such as:
     ```python
     def _find_triggered_methods(
         self, 
         trigger_method: str, 
         router_only: bool
     ) -> List[str]:
     ```

## tests/flow_test.py

### Positive Aspects
- **Test Coverage**: The tests provide comprehensive coverage of the new features.
- **Clarity in Test Cases**: Using meaningful names for test methods aids understanding at a glance.
- **Edge Case Testing**: The inclusion of edge cases improves robustness.

### Suggested Improvements

1. **Test Organization**:
   - **Current Implementation**: Some tests are too broad and encompass multiple scenarios distractingly.
   - **Improvement**: Decompose larger tests into smaller, focused units:
     ```python
     def test_router_execution_order(self, flow):
         # Focused test here
     ```

2. **Execution Tracker Fixture**:
   - **Current Implementation**: Duplicate state management could clutter tests.
   - **Improvement**: Incorporate fixtures to handle shared state uniformly:
     ```python
     @pytest.fixture(autouse=True)
     def execution_tracker():
         execution_order = []
     ```

3. **Error Case Testing**:
   - **Current Implementation**: Lack of tests for invalid states.
   - **Improvement**: Add tests addressing scenarios where routers yield invalid conditions:
     ```python
     def test_router_with_invalid_condition(self):
         # Specific logic for invalid conditions tested here
     ```

## General Recommendations
1. **Documentation**: Enhance codebase documentation, focusing on type hints and function purposes to aid future maintainers.
2. **Error Handling**: Add explicit error handling for failures encountered during router executions.
3. **Performance Benchmarking**: Introduce performance tests to monitor impacts of router chain execution times.
4. **Modular Design**: Consider refactoring and isolating complex logic into separate classes to enhance modularity.

## Conclusion
The modifications presented in PR #2231 succeed in enhancing the robustness of the flow execution logic, meeting the demands of the identified issue. The careful attention to code quality and testing standards is commendable. Incorporating the suggested improvements will further elevate the maintainability and performance of the codebase, ultimately leading to a more resilient application structure.

File: src/crewai/flow/flow.py [hunk 1]
```diff
@@ -894,35 +894,45 @@ async def _execute_listeners(self, trigger_method: str, result: Any) -> None:
...
-        )
-        if listeners_triggered:
-            tasks = [
-                self._execute_single_listener(listener_name, result)
-                for listener_name in listeners_triggered
-            ]
-            await asyncio.gather(*tasks)
+                router_result = self._method_outputs[-1]
+                if router_result:  # Only add non-None results
+                    router_results.append(router_result)
```
Comment (lorenzejay) [inline src/crewai/flow/flow.py:917]: should we break if None? or do we just need to know where to trigger next ?

**Review by lorenzejay — APPROVED** [review]

File: src/crewai/flow/flow.py (hunk 1 above)
Comment (bhancockio) [reply src/crewai/flow/flow.py:917]: https://www.loom.com/share/da216e85a94348dc8bda3b67d3837ccd?sid=90db863f-8fb2-4339-8d3a-5fdae098ef12
